x/mm2/keeper/keeper.go

```diff
@@ -0,0 +1,181 @@
+// The Ticker.String corresponds to a market, and must be unique.
+func (k *Keeper) createMarket(ctx sdk.Context, market types.Market) error {
+	// Check if Ticker already exists for the provider
+	alreadyExists, err := k.markets.Has(ctx, types.TickerString(market.Ticker.String()))
+	if err != nil {
+		return err
+	}
+	if alreadyExists {
+		return types.NewMarketAlreadyExistsError(types.TickerString(market.Ticker.String()))
+	}
+	// Create the config
+	return k.markets.Set(ctx, types.TickerString(market.Ticker.String()), market)
+}
+
+// updateMarket updates a Market.
+// The Ticker.String corresponds to a market, and exist unique.
+func (k *Keeper) updateMarket(ctx sdk.Context, market types.Market) error {
+	// Check if Ticker already exists for the provider
+	alreadyExists, err := k.markets.Has(ctx, types.TickerString(market.Ticker.String()))
+	if err != nil {
+		return err
+	}
+	if !alreadyExists {
+		return types.NewMarketDoesNotExistsError(types.TickerString(market.Ticker.String()))
+	}
+	// Create the config
+	return k.markets.Set(ctx, types.TickerString(market.Ticker.String()), market)
+}
+
+// CreateMarket sets the ticker, paths, and providers for a given market.  It also
+// sets the LastUpdated field to the current block height.
+func (k *Keeper) CreateMarket(ctx sdk.Context, market types.Market) error {
+	if err := k.createMarket(ctx, market); err != nil {
+		return err
+	}
+
+	return k.SetLastUpdated(ctx, uint64(ctx.BlockHeight()))
+}
+
+// UpdateMarket updaters the ticker, paths, and providers for a given market.  It also
+// sets the LastUpdated field to the current block height.
+func (k *Keeper) UpdateMarket(ctx sdk.Context, market types.Market) error {
+	if err := k.updateMarket(ctx, market); err != nil {
+		return err
+	}
+
+	return k.SetLastUpdated(ctx, uint64(ctx.BlockHeight()))
+}
+
+// SetParams sets the x/marketmap module's parameters.
+func (k *Keeper) SetParams(ctx sdk.Context, params types.Params) error {
+	return k.params.Set(ctx, params)
+}
+
+// GetParams returns the x/marketmap module's parameters.
+func (k *Keeper) GetParams(ctx sdk.Context) (types.Params, error) {
+	return k.params.Get(ctx)
+}
+
+// ValidateState is called after keeper modifications have been made to the market map to verify that
+// the aggregate of all updates has led to a valid state.
+func (k *Keeper) ValidateState(ctx sdk.Context, creates []types.Market, updates []types.Market) error {
+	for _, market := range append(creates, updates...) {
+		// check that all paths already exist in the keeper store:
+		for _, providerConfig := range market.ProviderConfigs {
+			if providerConfig.NormalizeByPair != nil {
+				has, err := k.markets.Has(ctx, types.TickerString(providerConfig.NormalizeByPair.String()))
+				if err != nil {
+					return err
+				}
+
+				if !has {
+					return fmt.Errorf("currency pair %s in provider config does not exist", providerConfig.NormalizeByPair.String())
+				}
+			}
+		}
+	}
+
+	return nil
+}
```

Comment:
I think this function might be a bit of a footgun since any external dependency to Create or Update will have to call validate state afterwards. Do you think it makes more sense to just move this into the top level Create and Update instead and have this be a internal helper function?